### ISSUE TITLE
stack negative value not show

### ISSUE THREAD

**Issue by godcanfly** [grafana/grafana#51011]
**What happened**:
stack negative value not show, according to https://github.com/grafana/grafana/issues/1360 fix and some fieled override method, this only work if the series data is all negative, but not the one has both negative and postive value.

I believe this is very basic function, I just cann't believe this still not working 

**What you expected to happen**:
stack according to the value
postion value stack above the zero line
negative value stack below the zero line 

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Grafana version:
- I have tried 8.2.0,8,5.1,9.00 all has the same issue
- Data source type & version:
- OS Grafana is installed on:
- User OS & Browser:
- Grafana plugins:
- Others:
![image](https://user-images.githubusercontent.com/17425165/174245348-68a7571d-57f2-4eee-9569-05d4ec2e1d39.png)

**Comment by gelicia** [comment]
@godcanfly can you please provide your dashboard JSON to help us replicate the issue?

**Comment by leeoniya** [comment]
this one will be difficult to address due to the fact that all renderers use the same high-performance fill-between-series strategy. only bar rendering style can even try to accommodate this if we decided to rewrite just that renderer for this case. this case of mixed pos/neg stacking is a relatively rare requirement to justify such an effort, imo, but we can leave this issue open for a while to see how many upvotes it gets. in either case, it is unlikely to happen soon.

**Comment by godcanfly** [comment]
> this one will be difficult to address due to the fact that all renderers use the same high-performance stacking code and fill-between-series strategy. only bar rendering style can even try to accommodate this if we decided to rewrite just that renderer for this case. mixed pos/neg stacking is a relatively rare requirement to justify such an effort, imo, but we can leave this issue open for a while to see how many upvotes it gets. in either case, it is unlikely to happen soon.

Thanks for the reply. I believe this is quite basic feature when handle both positive and negative, you don't expect them to be added together, they should be group separately to show since they increase in different situation. I believe the old issue https://github.com/grafana/grafana/issues/1360 should raise up the same requirement.

If this has to be the way. How can I group the data  accordion to the value instead of the field name? is there a way to do that?

**Comment by leeoniya** [comment]
> If this has to be the way. How can I group the data accordion to the value instead of the field name? is there a way to do that?

i dont think there is currently a way to do this, but that's an interesting work-around maybe. something like a transformer that can split single fields into +A and -A values fields. then you can set the same colors to both and each will be part of the + or -- stack separately.

**Comment by martian-maniac** [comment]
I am also experiencing this issue for Time series graph. I am graphing electricity grid data and some interconnecting cables switch between import/export.

Here is an example for PL expected to look like this:
![image](https://user-images.githubusercontent.com/96312338/185785222-2193db02-f082-428f-bde6-22e3728ca67e.png)

But when stacked with only-negative FI no positive PL values appear. It seems to cut up into FI instead.
![image](https://user-images.githubusercontent.com/96312338/185785243-36261554-cb91-4ce9-9214-f4d31aaeabce.png)

Source: https://intermittent.energy/d/_Zn37XmVz/sweden-import-export?orgId=1&from=1638831600000&to=1638918000000

**Comment by leeoniya** [comment]
i dont think mixed pos/neg stacking is generalizable for a line/area chart even if you can imagine how it might work in specific cases (e.g. only two series where one never crosses 0). most of the mixed stacking is with bars because you there is no line/fill to be drawn through the 0 crossover points in the middle of a stack.

https://www.google.com/search?q=positive+negative+stacking+chart&tbm=isch

**Comment by martian-maniac** [comment]
I don't think bar charts will be appropriate for high resolution data. I want to view one year of data and see hourly variations day by day. As workaround I will have to split import/export into separate series but then classic palette colors will be different which is not ideal. Assigning colors manually will be difficult as dataset includes all interconnects for a whole continent.

**Comment by leeoniya** [comment]
> I don't think bar charts will be appropriate for high resolution data.

agreed, but i have never seen what a mixed pos/neg area chart would look like, and am struggling to imagine it. do you have an example that shows one?

**Comment by martian-maniac** [comment]
I'm imagining it would just jump to the other side?
I hacked a query a bit for an example and, hmm. Area charts never jump it needs to connect line. Maybe transition to/from zero between datapoints? (I'm much better at data than visualization lol)

![image](https://user-images.githubusercontent.com/96312338/186135299-34d89641-74f6-4ed6-b95c-4214b169af2a.png)

**Comment by fgnn** [comment]
Another example of my use case. The trend below shows various components of electricity rates. The different tariffs can alternate between positive or negative month to month, but will always sum to a positive value for the overall rate (not all tariffs shown in the screenshot).

Single series with mixed positive and negative.
![image](https://user-images.githubusercontent.com/12035623/206789836-2364cd82-6799-4dda-8e58-cbb5fd0a3f28.png)

Enabled multiple series. Not only does the stacking not work as expected, but several of the bars have completely vanished (e.g. yellow).
![image](https://user-images.githubusercontent.com/12035623/206789906-11ea89bb-6fa8-46cb-b1d0-a80dad9794c4.png)

If I turn off stacked series, all the bars show up, but are overlapping each-other.
![image](https://user-images.githubusercontent.com/12035623/206790202-41cde638-6081-4012-83ca-746870a5cbe6.png)

I understand there are some of the technical limitations to solve this issue, but currently the visualization is quite corrupted with stacking mixed sign values.

Below is the panel with sample data. Make sure to change the timeline to years 2014-2018.

`{
  "id": 13,
  "gridPos": {
    "x": 0,
    "y": 0,
    "w": 12,
    "h": 8
  },
  "type": "timeseries",
  "title": "Panel Title",
  "datasource": {
    "uid": "9cY0WtPMz",
    "type": "testdata"
  },
  "pluginVersion": "9.3.2-45041",
  "targets": [
    {
      "scenarioId": "csv_content",
      "refId": "A",
      "datasource": {
        "type": "testdata",
        "uid": "9cY0WtPMz"
      },
      "csvContent": "Time,Balancing Pool Allocation Rider G,Distribution True Up Rider DJ,Transmission Deferral Rider K,Transmission True-Up Rider J\n1.41212E+12,-0.569,0.527,-0.424,-0.146\n1.4148E+12,-0.569,0.527,-0.424,0.109\n1.41739E+12,-0.569,0.527,-0.424,0.109\n1.42007E+12,-0.569,0.527,0.085,0.109\n1.42275E+12,-0.569,0.527,0.085,0.109\n1.42517E+12,-0.569,0.527,0.085,0.109\n1.42785E+12,-0.569,0.527,-0.387,0.109\n1.43044E+12,-0.569,0.527,-0.387,0.109\n1.43312E+12,-0.569,0.527,-0.387,0.109\n1.43571E+12,-0.569,0.527,-0.269,0.109\n1.43839E+12,-0.569,,-0.269,0.109\n1.44107E+12,-0.569,,-0.269,0.109\n1.44366E+12,-0.569,,0.166,0.109\n1.44634E+12,-0.569,,0.166,\n1.44893E+12,-0.569,,0.166,\n1.45161E+12,-0.336,-0.126,0.05,0.04\n1.45428E+12,-0.336,-0.126,0.05,0.04\n1.45679E+12,-0.336,-0.126,0.05,0.04\n1.45947E+12,-0.336,,0.255,\n1.46206E+12,-0.336,,0.255,\n1.46474E+12,-0.336,,0.255,\n1.46733E+12,-0.336,,0.285,-0.483\n1.47001E+12,-0.336,,0.285,-0.483\n1.47269E+12,-0.336,,0.285,-0.483\n1.47528E+12,-0.336,,-0.776,\n1.47796E+12,-0.336,,-0.776,-0.284\n1.48055E+12,-0.336,,-0.776,-0.284\n1.48323E+12,0.114,,-0.003,-0.594\n1.48591E+12,0.114,,-0.003,-0.594\n1.48833E+12,0.114,,-0.003,-0.594\n1.491E+12,0.114,-0.289,0.174,0.216\n1.4936E+12,0.114,-0.289,0.174,0.216\n1.49628E+12,0.114,-0.289,0.174,0.216\n1.49887E+12,0.114,,-0.202,0.216"
    }
  ],
  "options": {
    "tooltip": {
      "mode": "single",
      "sort": "none"
    },
    "legend": {
      "showLegend": true,
      "displayMode": "list",
      "placement": "bottom",
      "calcs": []
    }
  },
  "fieldConfig": {
    "defaults": {
      "custom": {
        "drawStyle": "bars",
        "lineInterpolation": "linear",
        "barAlignment": 0,
        "lineWidth": 1,
        "fillOpacity": 44,
        "gradientMode": "none",
        "spanNulls": false,
        "showPoints": "auto",
        "pointSize": 5,
        "stacking": {
          "mode": "normal",
          "group": "A"
        },
        "axisPlacement": "auto",
        "axisLabel": "",
        "axisColorMode": "text",
        "scaleDistribution": {
          "type": "linear"
        },
        "axisCenteredZero": false,
        "hideFrom": {
          "tooltip": false,
          "viz": false,
          "legend": false
        },
        "thresholdsStyle": {
          "mode": "off"
        }
      },
      "color": {
        "mode": "palette-classic"
      },
      "mappings": [],
      "thresholds": {
        "mode": "absolute",
        "steps": [
          {
            "value": null,
            "color": "green"
          },
          {
            "value": 80,
            "color": "red"
          }
        ]
      }
    },
    "overrides": []
  }
}`

**Comment by amarant** [comment]
It seems the bar chart panel share the stacking code with time series panel, but if I understand the complexity and originality of mixed positive/negative series in stacked graph it is not unheard of :

![image](https://pbs.twimg.com/media/FIz7WiTWQAANfb2?format=jpg&name=large)


But in the bar chart world this is pretty common and all library work by default :

Here using eCharts VS the default bar chart where series overlaps :

![image](https://user-images.githubusercontent.com/2281471/217835837-21b3c2d8-3d26-402c-88db-06a906f2b17d.png)

**Comment by morganchristiansson** [comment]
This problem is widespread. Here's an example from http://www.caiso.com/TodaysOutlook/Pages/supply.html (California electric grid operator).  Enable stacking in options dropdown it's off by default.

Hydro
![image](https://github.com/grafana/grafana/assets/66844/c5c6bf9f-c3b4-44e7-b01c-a78c481df236)
\+ Batteries
![image](https://github.com/grafana/grafana/assets/66844/c8bfd201-c18e-47dc-ba4d-0f140b159427)
= Wrong
![image](https://github.com/grafana/grafana/assets/66844/9b9f77cc-2514-4b46-b3dd-4e91b815021f)

If anyone has examples of graphs showing mixed pos/neg data correctly I think it would be helpful.

**Comment by tegusi** [comment]
Meet the same problem waiting to get solution later

**Comment by intermittentnrg** [comment]
Here is an example of a correct stacked area chart.

The area jumps between pos/neg side and is drawn in the same color, as I suggested in https://github.com/grafana/grafana/issues/51011#issuecomment-1223873883

![bild](https://github.com/grafana/grafana/assets/96312338/70e6a839-db34-4346-925e-6dadebd292c7)

**Comment by leeoniya** [comment]
so, here's what i got as a test (fully opaque for now):

```js
let data = [
  [0,1,2, 3,4,5, 6, 7, 8,9], // x
  [1,1,1,-1,1,1, 1,-3, 1,1], // y1 (red)
  [1,1,1, 1,0,0,-1,-2,-3,2], // y2 (blue)
];
```

does this look like what you'd expect? (zero y is in middle, zero x is on left)

<img width="385" alt="image" src="https://github.com/grafana/grafana/assets/43234/29561aa1-10d4-45e0-a786-ff84a28de1a4">

**Comment by leeoniya** [comment]
here is how LibreOffice does mixed stacking. the bar versions are logical, and the area versions are a bunch of inscrutable nonsense: each sequential series is stacked away from the accumulated value at each x, which makes sense only on paper if the intent is to do a "sum" but when rendered literally, it ceases to make sense. e.g. in the bottom right chart, the sum total at x=7 should be 0 (presumable read off the down-stacked -1 red at that point? at x=4 the sum should be 0, red=-1, blue=1. how is anyone supposed to interpret this viz at all? :exploding_head: 

<img width="1417" alt="image" src="https://github.com/grafana/grafana/assets/43234/adcc933f-9960-4fba-8845-61190d45ef8a">

**Comment by fheck** [comment]
@leeoniya 
The visual in the LibreOffice charts is a little off, I give you that, but your visual is plainly wrong. Compare x=5 in the right visual to yours. Without knowing the values, but knowing the graph is stacked "blue" being 0 is not a possible conclusion in your visual.

You say 
> when the intent is to do a "sum"

as if that would be an afterthought. But a stacked graph does precisely that, it shows the sum when there is a cumulative property about the different categories. So the sum being correctly represented in the height of the graph is a fundamental goal. 
I think the order of the series plays a vital role in the interpretation. Since blue is the "upper" series, the "sum" in your example x=7 is correctly shown in the LibreOffice graph, albeit ambiguous: the interpretation coming to the correct conclusion would be "red"=+1 which means blue needs to start at +1, now since "blue"=-1, and blue is the "upper" series, the "lower" edge of the blue series is at +1 and the "upper" at 0, fully blocking any red area. 
As I said, there is another interpretation which can be gotten from the graph: red=0 and blue=1. But at least the sum and the ordering is consistent, which is not the case in your rendering.

I'm in no way an expert, I just subscribed to this issue because, as may others here, I tried to graph energy yields from a hybrid solar system with a battery. And I found that in some cases the current bar chart would be able to show an intuitive rendering, but in other situations it completely fell apart on the same data. So I assume besides the conceptional problems raised in this thread, there must also be an actual bug in the current implementation. 

And as you observed there might be a difference between bars and lines/areas. So while I appreciate the fact that the implementation is shared between the two, it might not be possible to have a "one size fits all" logic in this case. The initial bug report *is*, after all, about the bar chart not the line chart.

**Comment by leeoniya** [comment]
> The visual in the LibreOffice charts is a little off, I give you that, but your visual is plainly wrong. Compare x=5 in the right visual to yours. Without knowing the values, but knowing the graph is stacked "blue" being 0 is not a possible conclusion in your visual.

yes, i considered this. there is a modification to my version that can be done to continue to "stack" 0 values onto their previous signed stack, so blue 0 at x=5 would continue straight across to x=6. this would inflate the volume of the blue fill in that x=5-6 "trench" which is a bit misleading with _this_ dataset, and would be _very_ misleading if there was a gap in time there. for example if that x delta was a "day" wide, but previous deltas were "hour" wide. you would then have a huge trench of filled blue that was actually 0 for the whole duration. (i can show you what that looks like in a bit).

> as if that would be an afterthought. But a stacked graph does precisely that, it shows the sum when there is a cumulative property about the different categories. So the sum being correctly represented in the height of the graph is a fundamental goal.

yes, okay. so your strategy is to read the sum off the top of the "blue" stack? that works great at x=4 and fails great at x=7, x=9, etc.

<img width="1026" alt="image" src="https://github.com/grafana/grafana/assets/43234/7b71793a-e4d9-4cd4-b7ec-bf00d1b15c50">

**Comment by leeoniya** [comment]
if we set the opacity of each to 50%, then it becomes very slightly less bad:

<img width="459" alt="image" src="https://github.com/grafana/grafana/assets/43234/ad50b302-f603-4b52-bf4e-2858901cb7da">

**Comment by leeoniya** [comment]
this is what it looks like if you stack 0. you can see at x=4 now the entire "trough" is filled with blue. i think this is misleading even in this narrow x and small y case.

https://jsfiddle.net/qzdxbp5r/

<img width="384" alt="image" src="https://github.com/grafana/grafana/assets/43234/a09200dd-75f0-4c06-9562-41b4db04c7bf">

---

if we change the data slightly to increase the gap between x=4 and x=5, and make the red value at x=5 let's say 4 instead of 1:

https://jsfiddle.net/u1jnzo3v/

<img width="608" alt="image" src="https://github.com/grafana/grafana/assets/43234/6e37b101-448b-4a70-8fd0-b9721d34c015">

so, this is where "blue" is actually zero:

<img width="185" alt="image" src="https://github.com/grafana/grafana/assets/43234/dc801df2-1b5b-4bb4-ba62-8eba8fb680fe">

here is the same thing without zero stacking. it's better, but still sucks:

<img width="607" alt="image" src="https://github.com/grafana/grafana/assets/43234/eca7a659-f2c0-45fa-aaf4-131aa4d6dab9">

**Comment by fheck** [comment]
Nothing has changed, the bar plot is still broken when using stacked series when there are negative values. So yes, it is still relevant.
